Libs/MRML/Core/vtkMRMLModelNode.cxx

```diff
@@ -141,64 +173,130 @@ void vtkMRMLModelNode::SetAndObservePolyData(vtkPolyData *polyData)
+    {
+    return polydata;
+    }
+  else
+    {
+    this->GeometryFilter->Update();
+    return this->GeometryFilter->GetOutput();
+    }
+}
+
+//---------------------------------------------------------------------------
+vtkUnstructuredGrid* vtkMRMLModelNode::GetUnstructuredGrid()
+{
+  return vtkUnstructuredGrid::SafeDownCast(this->GetMesh());
 }
 
 //---------------------------------------------------------------------------
 void vtkMRMLModelNode
-::SetPolyDataConnection(vtkAlgorithmOutput *newPolyDataConnection)
+::SetMeshConnection(vtkAlgorithmOutput *newMeshConnection)
 {
-  if (newPolyDataConnection == this->PolyDataConnection)
+  if (newMeshConnection == this->MeshConnection)
     {
     return;
     }
 
-  vtkAlgorithm* oldPolyDataAlgorithm = this->PolyDataConnection ?
-    this->PolyDataConnection->GetProducer() : 0;
+  vtkAlgorithm* oldMeshAlgorithm = this->MeshConnection ?
+    this->MeshConnection->GetProducer() : 0;
 
-  this->PolyDataConnection = newPolyDataConnection;
+  this->MeshConnection = newMeshConnection;
 
-  vtkAlgorithm* polyDataAlgorithm = this->PolyDataConnection ?
-    this->PolyDataConnection->GetProducer() : 0;
-  if (polyDataAlgorithm != NULL)
+  vtkAlgorithm* newMeshAlgorithm = this->MeshConnection ?
+    this->MeshConnection->GetProducer() : 0;
+  if (newMeshAlgorithm != NULL)
     {
     vtkEventBroker::GetInstance()->AddObservation(
-      polyDataAlgorithm, vtkCommand::ModifiedEvent, this, this->MRMLCallbackCommand );
-    polyDataAlgorithm->Register(this);
+      newMeshAlgorithm, vtkCommand::ModifiedEvent, this, this->MRMLCallbackCommand );
+    newMeshAlgorithm->Register(this);
     }
 
-  this->SetPolyDataToDisplayNodes();
+  this->SetMeshToDisplayNodes();
 
-  if (oldPolyDataAlgorithm != NULL)
+  if (oldMeshAlgorithm != NULL)
     {
     vtkEventBroker::GetInstance()->RemoveObservations (
-      oldPolyDataAlgorithm, vtkCommand::ModifiedEvent, this, this->MRMLCallbackCommand );
-    oldPolyDataAlgorithm->UnRegister(this);
+      oldMeshAlgorithm, vtkCommand::ModifiedEvent, this, this->MRMLCallbackCommand );
+    oldMeshAlgorithm->UnRegister(this);
     }
 
   this->StorableModifiedTime.Modified();
   this->Modified();
-  this->InvokeEvent( vtkMRMLModelNode::PolyDataModifiedEvent , this);
+  this->InvokeEvent( vtkMRMLModelNode::MeshModifiedEvent , this);
+}
+
+//---------------------------------------------------------------------------
+void vtkMRMLModelNode
+::SetPolyDataConnection(vtkAlgorithmOutput *newPolyDataConnection)
+{
+  this->MeshType = vtkMRMLModelNode::PolyDataMeshType;
+  this->SetMeshConnection(newPolyDataConnection);
+  this->GeometryFilter->SetInputConnection(0);
+}
+
+//---------------------------------------------------------------------------
+void vtkMRMLModelNode
+::SetUnstructuredGridConnection(vtkAlgorithmOutput *newUnstructuredGridConnection)
+{
+  this->MeshType = vtkMRMLModelNode::UnstructuredGridMeshType;
```

Comment:
Difficult to tell what's the best solution. Probably keeping just the two actual types is simpler (less chance for inconsistency) and by default assuming that we have NULL polydata int he node is more backward-compatible. So, unless you find a use case that's actually broken with the current implementation, I would keep things as they are now. My comment was intended just for giving some thoughts and testing use cases around reading/writing of the node. For example, make sure that there is no crash if the mesh file is not available (it was deleted, etc); or it's the wrong type.